cmd/mirror-url.go

```diff
@@ -113,9 +114,15 @@ func deltaSourceTarget(sourceURL, targetURL string, isFake, isOverwrite, isRemov
 
 	// List both source and target, compare and return values through channel.
 	for diffMsg := range objectDifference(sourceClnt, targetClnt, sourceURL, targetURL) {
+
 		if diffMsg.Error != nil {
+			// End the process if connection is refused,
+			// otherwise ignore the error and proceed to next object.
+			if _, ok := diffMsg.Error.ToGoError().(*url.Error); ok {
```

Comment:
Good catch @ebozduman !

However, I don't imagine why we are only erroring out network connection issues and why we *were* ignoring errors in the first place.

objectDifference() is listing objects in source & target and then calculating the difference, I don't see any possible errors that can be ignored in this case.

I think we should return for all errors and fix later if users eventually complain.

Reply:
I think that is about how watch logic works with `mc mirror`.

The requirement/design is such that we don't stop when there is an error and keep continue with the rest of the source file list until the list is exhausted. Like, if there are permission problems associated with the source or target files, and in error conditions like `TooManySymLinks`, `PathNotFound`, etc. we want to continue to complete going through the full source list. 
However, there are fatal errors we should not ignore.
So, when there is a fatal error like network connection problem with the source server, we kept going instead of stopping the mirroring process. Any other error condition, in which we can communicate with both source and target, we should keep continue.

Reply:
yeah, you are right, I forgot about these FS errors in listing.

But also, S3 listing can return some errors. S3 returns list of objects 1000 by 1000, if S3 decides to return 500 status code or SlowDown, etc.. then we will have the same problem, and objects can be deleted without the user's consensus.

I really recommend that we do the reverse way, we decide (not necessarily now) which errors we can safely ignore and return all other errors. What do you think @ebozduman ?

Reply:
Sure, that might be possible.
Let me also discuss this with @harshavardhana

Reply:
@vadmeste,

I think what you suggest depends on which error condition list, ignore list or error out list, we can come up fully with. In other words, we need to know the exact list of errors we'll either ignore and remove files or display and halt the action when hit. 

After discussing with @harshavardhana this morning about how we can decide to error out when `remove` action is requested, he came up with a different suggestion, which I'll share with you to get your opinion and then also discuss with @abperiasamy.

Coming out with a **full** list of error messages in both cases is not easy, so it is most likely that the file list we choose to come up with is not going to be complete. Besides, the after affect for `copy`/`overwrite`, and `remove` transactions are drastically different, when we hit an error condition, and the returned file list is erratic. That is; if we make the wrong decision and remove some files from the target by mistake, the target file/data is lost, while when some files are NOT copied over or overwritten, we'll still be fine.

With that in mind, @harshavardhana suggested to always stop when there is any kind of errors regardless of its type with the claim that the produced file list cannot be trusted when there is an error and hence we should not proceed with any action. Since we'll display the error message in this scenario, user will have an idea about the reason why mirroring did not happen. 
With this design change, we'll be safe in `remove` since no action has been taken.  With `copy/overwrite`, mirroring will just be delayed.

Let me know what you think about the above suggestion, that is; if we always error out instead of trying to make intelligent decisions to be able to selectively choose among error conditions.
Please also let me know if there is a need for more clarification.

Reply:
@ebozduman yeah I agree with this solution as a first step. If someday, some users complain and say that mc  should ignore some of them we can take a look again.

Reply:
I expect that we would be to construct a list of errors that we can safely ignore. But let's start by making mirroring stops for every error.

Reply:
Actually my mistake we need to bring in errors that we need to ignore. Because some objects can be on `Glacier` and we need to skip them.

Reply:
@harshavardhana, @vadmeste,
We currently ignore the following errors:
1. BrokenSymlink, 
2. TooManyLevelsSymlink, 
3. PathNotFound,
4. PathInsufficientPermission, 
5. ObjectOnGlacier

We stop on all other errors.

I've tested this fix in the following scenarios:
1. CTRL-C'ed source server and saw the following error message. Then restarted `mc mirror --remove --watch` and confirmed that the target bucket content is not removed as  it was happening before this fix.
2. I've also tested the fix by having a file with no permissions and mirroring did not stop when it hit the permission problem.

If you think there are more errors we need to be aware of and ignore, and if you can think of other tests that can be run to verify the fix, please let me know.